File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
// Use Scanner

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Scanner;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {

        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int M = sc.nextInt();

        int[] tree = new int[N];

        int min = 0;
        int max = 0;


        for (int i = 0; i < N; i++) {
            tree[i] = sc.nextInt();
        }
        Arrays.sort(tree);

        max = tree[N - 1];

        // 이진 검색(Upper Bound)
        while (min < max) {

            int mid = (max + min) / 2;
            long sum = 0;

            for (int treeHeight : tree) {
                if (treeHeight - mid > 0) {
                    sum += (treeHeight - mid);
                }
            }

            if (sum < M) {
                max = mid;
            } else {
                min = mid + 1;
            }
        }
        System.out.println(min - 1);
    }
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
from sys import stdin

val = list(map(int, stdin.readline().split()))

N = val[0]
M = val[1]

tree = sorted(list(map(int, stdin.readline().split())))

min = 0
max = 0

max = tree[N-1]

while(min < max):
    mid = (max + min) // 2
    sum = 0

    for treeHeight in tree:
        cutTree = treeHeight - mid
        
        if(cutTree > 0):
            sum += cutTree
    
    if(sum < M):
        max = mid
    else:
        min = mid + 1

print(min - 1)
